### ISSUE TITLE
morphocluster UI for quality control in FathomNet

### ISSUE THREAD

**Issue by kakanikatija** [fathomnet/community-feedback#17]
Good to implement for rapid QC of FathomNet submissions: https://arxiv.org/abs/2005.01595

**Comment by hohonuuli** [comment]
Morphocluster source code lives at <https://github.com/morphocluster/morphocluster>. @kevinsbarnard was able to get it up and running quickly following the instructions in the repo. The default setup doesn't do a great job of clustering for our images, maybe use one of our models (Just strip out the classifier labels). @moi90 recommended keeping the features to 32 dimensions, more than that runs sloooooowly. 

Can we integrate this into gridview to group like images near each other?

**Comment by hohonuuli** [comment]
Note: That morphocluster only handles one dataset at a time. Usage currently requires spinning up a new instance for each dataset.

**Comment by hohonuuli** [comment]
__TODO__: Run tests with Morphocluster using our own feature extractor

**Comment by hohonuuli** [comment]
50k images is a low end for Morphocluster, it's more optimized for sets on the order of 5 million.

**Comment by moi90** [comment]
Regarding the clustering of the color images:
- As discussed, a very promising approach would be to use the features of the object detector model. Alternatively, one could use a model trained on ImageNet. A more advanced approach could be the autoencoder idea.
- For now, I would try to decrease the min_cluster_size to see if you are able to get more homogeneous clusters this way.

**Comment by hohonuuli** [comment]
@kevinsbarnard ^^^^ When you get a minute would you try @moi90 suggestion? Thanks!

**Comment by kevinsbarnard** [comment]
@hohonuuli Just ran clustering with a min cluster size of 128, 64, 32, 16, and 8 on the same set of 4758 midwater ROIs from before. I'm not sure how to quantify it, but at a glance the initial clusters are looking more homogeneous, and of course there are a lot more of them to validate. I'd be curious to try it out on benthic images as well.

![Screenshot_20220126_120453](https://user-images.githubusercontent.com/40082734/151238644-4ca6b617-e751-47b8-aa22-378b30547021.png)

**Comment by lonnylundsten** [comment]
Cool. It looks good to me. I'd love to see it on some of the benthic sets.

**Comment by kevinsbarnard** [comment]
@lonnylundsten @hohonuuli I generated a 2-class benthic set (via m3-download) consisting of 659 ROIs total: a mix of _Chionoecetes tanneri_ and _Paragorgia arborea_ (picked because a non-expert like myself can still discern). With a min cluster size of 32, three clusters were generated, which look promising already:
![Screenshot_20220126_134712](https://user-images.githubusercontent.com/40082734/151254219-9cf6c3c2-0a84-49c7-ab08-5f6eabd92b86.png)
![Screenshot_20220126_134759](https://user-images.githubusercontent.com/40082734/151254255-90195427-0e4e-418e-92ed-8d3aa00aa366.png)
![Screenshot_20220126_134828](https://user-images.githubusercontent.com/40082734/151254270-93e4585e-4fbb-4894-a75c-29c7808b5e0e.png)
It took me about 10 minutes to do the validate and grow steps for the first iteration. For one cluster, the suggestions were evenly mixed, and for the other two they were nearly perfect.

**Comment by lonnylundsten** [comment]
Looks good, Kevin. So that’s just the morphocluster code? That’s not using the feature extractor from the yolo model? I’d like to have you walk me through the interface because It’s not super clear, for example if a cluster contains a mix of classes if there mechanism to remove a bad class from a cluster and move it to the proper cluster? I do see how this can be useful but I’m curious about that human interaction.

**Comment by kevinsbarnard** [comment]
@lonnylundsten Yes, these were generated using the feature extractor provided with the morphocluster repository. In the validation stage (shown in those screenshots), you can kick an ROI out of the cluster by pressing the up arrow at the top right (moves it into the parent node).

**Comment by moi90** [comment]
@lonnylundsten In order to maximize efficiency, the preferred way is not to resort the objects of inhomogeneous clusters, but to delete the cluster altogether. The objects of the inhomogeneous clusters will eventually reappear as clusters in a later iteration of the process. Or maybe the correct cluster for these objects already exists and the objects are pulled in during growing.

One tweak that might make sense to implement: If a cluster is clearly two categories, it could be split automatically. The resulting clusters would then be validated as usual.

**Comment by hohonuuli** [comment]
@kakanikatija suggests adding a gridview-like tool to FathomNet for rapid assesment/correction.

**Comment by Jordan-Pierce** [comment]
This is awesome! Have you looked at other unsupervised/self-supervised techniques other than MorphCluster (which also looks dope)?

[SimCLR ](https://ai.googleblog.com/2020/04/advancing-self-supervised-and-semi.html)is currently one of the SOTA methods.

**Comment by moi90** [comment]
Self-Supervised feature extractor training is certainly the way to go for new datasets and will increase MorphoCluster's efficiency. But in itself, it doesn't help you with annotating ;)
